src/asm/lexer.c

```diff
@@ -88,17 +91,43 @@ void yyunput(char c)
 	*(--pLexBuffer) = c;
 }
 
-void yyunputstr(char *s)
+void yyunputstr(const char *s)
 {
-	int32_t i, len;
+	int32_t len;
 
 	len = strlen(s);
 
-	if (pLexBuffer - len < pLexBufferRealStart)
+	pLexBuffer -= len;
+
+	if (pLexBuffer < pLexBufferRealStart)
```

Comment:
Both the original code and the new code can result in undefined behavior. From C11 6.5.6/8:

> When  an  expression  that  has  integer  type  is  added  to  or  subtracted  from  a  pointer,  the result has the type of the pointer operand. ... If  both  the  pointer operand  and  the  result  point  to  elements  of  the  same  array  object,  or  one  past  the  last element of the array object, the evaluation shall not produce an overflow; otherwise, the behavior is undefined.

Either `pLexBuffer - len` or `pLexBuffer -= len` could result in a pointer that points to something before the beginning of the array. The check has to be done before this occurs.

```
if (pLexBuffer - pLexBufferRealStart < len)
	fatalerror("Buffer safety margin exceeded");

pLexBuffer -= len;
```

Reply:
Thank you for pointing that out, fixed it in 476ccc9